ols/utils/auth_dependency.py

```diff
@@ -0,0 +1,210 @@
+    """
+
+    _instance = None
+    _api_client = None
+    _auth_api = None
+    _authorization_api = None
+
+    def __new__(cls):
+        """Create a new instance of the singleton, or returns the existing instance.
+
+        This method initializes the Kubernetes API clients the first time it is called.
+        and ensures that subsequent calls return the same instance.
+        """
+        if cls._instance is None:
+            cls._instance = super(K8sClientSingleton, cls).__new__(cls)
+            configuration = kubernetes.client.Configuration()
+
+            try:
+                try:
+                    kubernetes.config.load_kube_config(
+                        client_configuration=configuration
+                    )
+                except Exception:
+                    try:
+                        logger.warn(
+                            "kubeconfig file not found, trying in-cluster configuration"
+                        )
+                        kubernetes.config.load_incluster_config(
+                            client_configuration=configuration
+                        )
+                    except Exception:
+                        logger.warn(
+                            "Failed to load in-cluster Kubernetes configuration"
+                        )
+                        logger.warn("Trying to load k8s config from app config")
+                        if (
+                            config.ols_config.authentication_config.k8s_cluster_api
+                            and config.dev_config.k8s_auth_token
+                        ):
+                            configuration.api_key["authorization"] = (
+                                config.dev_config.k8s_auth_token
+                            )
+                            configuration.api_key_prefix["authorization"] = "Bearer"
+                        else:
+                            logger.error(
+                                "Failed To load K8S Config, Cant initialize Client"
+                            )
+                            raise
+                configuration.host = (
+                    config.ols_config.authentication_config.k8s_cluster_api
+                    if config.ols_config.authentication_config.k8s_cluster_api
+                    else configuration.host
+                )
+                configuration.verify_ssl = (
+                    not config.ols_config.authentication_config.skip_tls_verification
+                )
+                configuration.ssl_ca_cert = (
+                    config.ols_config.authentication_config.k8s_ca_cert_path
+                    if config.ols_config.authentication_config.k8s_ca_cert_path
+                    else configuration.ssl_ca_cert
+                )
+                api_client = kubernetes.client.ApiClient(configuration)
+                cls._api_client = api_client
+                cls._auth_api = kubernetes.client.AuthenticationV1Api(api_client)
```

Comment:
`authentication` or `authn` if you want to keep it short (and then authz for authorization).  But just `auth` is ambiguous.

Reply:
changed to authn for authentication and authz for authorization